README.md
# T-Builder

Allows the user to plot any polyprotic or polyfunctional titration curve.

Run the executable to open the GUI.

Enter the data as indicated into the data entry area of the GUI. Click "Save Titration Data" or "Save Bjerrum Data" to
save the data plotted above to a CSV file.

## Features

* Plot Titration curves and Relative Species Plots
* Arbitrarily protic or functional analyte
* Arbitrarily protic or functional titrant
* Calculate Equivalence Points
* Calculate buffer points. Where pH = pKa or pKb
* View Derivatives of plots
* Save Titration or Relative Species plots to CSV

### Citric acid titrated with KOH:

![citric_acid_koh_titration](https://i.imgur.com/gQIjOxH.png)

### Species of Citric acid as pH increases:

![citric_acid_relative_speciation](https://i.imgur.com/Ng8gZpb.png)

## TODO:

* <del>"Show Plot" method</del> ***Done 7/31/2020***
* <del>"Save Data CSV" method</del> ***Done 7/31/2020***
* <del>Make a N-Functional analyte option</del> ***Done 12/23/2020***
* <del>Make a M-Functional titrant option</del> ***Done 12/23/2020***
* <del>Support for showing Bjerrum Plots </del> ***Done 12/23/2020***
* <del>EQ point finder (either with phi or with 2nd deriv.)</del> ***Done 12/28/2020***
* <del>Support for different temperatures </del> ***Done 12/24/2020***
* <del>A GUI</del> ***Done 12/26/2020***
* <del>Make a better name</del> ***Done 12/26/2020***
* <del>More Plot options</del>
    * <del>Probably just pass through matplotlib figure **kwargs</del> ***No longer pursued due to GUI***
* Figure out how to make the titration curve of a mixture of analytes
* Optional activity coefficients.
* Optional Species Names
* Advanced Features
    * <del>Calculate the equivalence points</del> ***Done 12/28/2020***
    * <del>Plot the derivatives</del> ***Done 12/28/2020***
    * Allow for preset compounds
    * Saving and loading compounds
* <del>In GUI support for different temperatures<\del> ***Done 1/2/2021***

## Math

Variables with the subscript 't' refer to the titrant. Variables with the subscript 'a' refer to the analyte.

Equations are a generalized form of the equations found
in [Quantitative Chemical Analysis 9th Ed.](https://www.amazon.com/Quantitative-Chemical-Analysis-Daniel-Harris/dp/146413538X)
by Daniel C. Harris. (Ch. 11.10)

The alpha value for a species tells the relative predominance of the species at a given pH. The values range between 0
and 1. The larger the alpha value for the species, the higher concentration the species is in at that pH. The alpha
value for a given species at a certain concentration of hydronium can be given by the equation:

![alpha_equation](https://latex.codecogs.com/png.latex?\dpi{200}&space;\bg_white&space;\fn_cm&space;\alpha_s&space;=&space;\frac{\Omega_s}{\sum_{n,m=\i,0}&space;^{0,&space;\i}([H^&plus;]^{n}&space;*&space;\prod^{m}&space;_{j=0}(K_j))})

where Omega_s is equal to the component of the sum in the denominator with the index equal to the absolute value of the
charge on the species (assuming the un-reacted species is neutral). The un-reacted species, the species with no charge,
s = 0, and thus Omega_s would equal the first component of the summation in the denominator. The value of i is the
degree of functionality for the analyte. K_j represents the indexed dissociation constant for the analyte (For this
notation, assume K_0 = 1). These values can be directly ported into a bjerrum plot or predominance zone diagram.

With the alpha values now calculated, another relationship can be utilized. The following equation can be obtained:

![phi_definition](https://latex.codecogs.com/png.latex?\dpi{200}&space;\bg_white&space;\phi&space;\equiv\frac{C_tV_t}{C_aV_a}=&space;\frac{(\sum_{n=0}^{\i}n\alpha_{an})\pm\frac{[H^&plus;]-[OH^-]}{C_a}}{(\sum_{m=0}^{\j}m\alpha_{tm})\mp\frac{[H^&plus;]-[OH^-]}{C_t}})

where phi is defined as the "Fraction of the way to the equivalence point", and i and j are the functionality of the
analyte and titrant, respectively, alpha_an represents the indexed alpha values of the analyte, and the similarly,
alpha_tm represents the indexed alpha values of the titrant. The ± in the numerator is positive if the analyte is a
base, and negative if the analyte is an acid. The opposite is true for the denominator, as indicated. C_a and V_a are
the concentration and volume of the analyte. C_t and V_t are the concentration and volume for the titrant.

Phi can be used to determine the volume of titrant required to reach a certain pH. Since all the values in the
definition of phi are known, they can be solved to find phi. Using phi and known values for the volume of analyte, and
the concentration of both analyte and titrant, the volume of titrant can be calculated through re-aranging the previous
equation to find:

![phi_usage](https://latex.codecogs.com/png.latex?\dpi{200}&space;\bg_white&space;V_t&space;=&space;\frac{\phi&space;C_a&space;V_a}&space;{C_t})

This equation is then solved thousands of times to produce a plot of volumes to pH values. The volume of titrant is set
as the x-axis, and the pH the y-axis, and a titration curve is born.



CHANGELOG.txt
T-Builder Changelog

A log of all changes which affect code performance.

-----------------------------------------------------------------------
Version 2.3.1
-----------------------------------------------------------------------

New:
- A large list of compounds to select from
- Type hints
- Option to omit file headers in the save_files functions
- Consolidated classes into two dataclasses


Fixed:
- Speciation numbers weren't totalling 1
   - They were being scaled by their index improperly
   - Never underestimate the mutability of lists

Updated:
- Updated to DearPyGUI v0.6.121
- Updated to scipy v1.6.0
- README.md function variables
- Temperature now only scales from 0 to 100, rather than 0 to 350.

Removed:
- File printing tests fro tests.py
- All the old data files


-----------------------------------------------------------------------
Version 2.3
-----------------------------------------------------------------------

New:
- Refactored / Redesigned Gui
    - Larger Plots
    - Temperature adjustment in GUI
    - Save analysis data button
- Testing suite so I can test class changes more quickly
    - Test functions
    - Test data to compare against
- Save analysis function created and integrated to GUI.
- GUI examples to the README.md
- Link to jkelowitt/t-builder on title button.


Fixed:
- Labelled the titration generation setup.
    - The volume input is never where I think it is. This brute forces
        the input data to be put in the right places. New test data
        has been generated.

Updated:
- Increased the speed of the scale_alphas method by an order of
    magnitude
- Simplified alpha value calculation
- Removed temperature range warning from get_kw function.


Removed:
-



-----------------------------------------------------------------------
Version 2.1
-----------------------------------------------------------------------

New:
- Author, version number, and license
- Precision input
- Code is now formatted to the black formatting standard

Fixed:
- Derivatives wouldn't work on basic solutions
    - This fix has the unintended benefit that the saved data files for
        basic solutions are now in the correct order. Since the basic
        solutions were calculated from 14 --> 0 instead of the normal
        0 --> 14, the data files were backwards and the derivatives
        broken.
- When the buffer point isn't exactly calculated due to precision errors
    the closest value will be chosen. This is true for the equivalence
    point.
- Concentrations can no longer be negative. Zero concentrations are
    calculated but not shown

Updated:
- Plots are now anti-aliased
- query_callback functions are now a single function
- Moved closest value calculating lambda function into a proper function
- DearPyGui --> Version 0.6.117
- Precision default kwargs
- Scaling function changed from sigmoid to algebraic (x / (1+|x|))


Removed:
- Unused function imports


-----------------------------------------------------------------------
Version 2.1
-----------------------------------------------------------------------

New:
- Icon
- requirements.txt file
- Equivalence point calculation
- Buffering point calculation
- Derivative calculations

Fixed:
- Various grammatical errors
- The Titration data values were being saved in the wrong order


Updated:
- trim_values function now allows the user to input any list as an arg
    to be trimmed to the correct length

Removed:
- Compounds file


-----------------------------------------------------------------------
Version 2.0
-----------------------------------------------------------------------

New:
- GUI  -->  DearPyGui
- Name! T-Builder

Updated:
- README.md instructions on how to run the program

Removed:
- Matplotlib
    - Since the plotting is taken care of in the GUI, this is no longer
       required.


-----------------------------------------------------------------------
Version 1.1
-----------------------------------------------------------------------

New:
- Pulled Bjerrum and AcidBase classes out of the Titration class
- Added compound classes (including acidic and basic water)
- Tiny library of compound classes

Fixed:
- A call before definition error with __init__ of the AcidBase class

Updated:
- Titration class now requires you to make two compound classes and
   pass them in with their concentrations and the volume of analyte.


-----------------------------------------------------------------------
Version 1.0
-----------------------------------------------------------------------

New:
- Updated to Python 3.8
- Titration class
- Equivalence volume calculator
- Alpha value curve plotting
- Example classes

Fixed:
- Ordering of write titration data columns
- All mathematical errors

Removed:
- All traces of GUI


T-Builder.py
from webbrowser import open

from dearpygui.core import *
from dearpygui.simple import *

import data_writing as dw
from titration_class import Compound, Titration

__author__ = "jkelowitt"
__version__ = "v2.3.4"
__license__ = "MIT"

plot_width = 905
plot_height = 755
data_width = 200


def open_link(sender, data):
    open("https://github.com/jkelowitt/t-builder", new=2)


def query(sender, data):
    set_plot_xlimits(sender, data[0], data[1])
    set_plot_ylimits(sender, data[2], data[3])


def make_titration():
    # Create compounds
    Analyte = Compound(
        name=get_value("Analyte Name"),
        acidic=get_value("aa"),
        pKas=[float(i) for i in get_value("apk").split(",")],
    )

    Titrant = Compound(
        name=get_value("tname"),
        acidic=not get_value("aa"),
        pKas=[float(i) for i in get_value("tpk").split(",")],
    )

    # Create titration object
    titr = Titration(
        analyte=Analyte,
        titrant=Titrant,
        concentration_analyte=get_value("aconc"),
        concentration_titrant=get_value("tconc"),
        volume_analyte=get_value("avol"),
        decimal_places=get_value("precision"),
        temp=get_value("temperature"),
    )

    return titr


def plot_callback(sender, data):
    clear_plot("Main Plot")

    # Band-aid fix to issue #10. Everything which needs to be kept in a certain range must have callback_data
    try:
        value = get_value(sender)
        if value < data[0]:
            set_value(sender, data[0])
            clear_plot("Main Plot")
            return
    except:  # If the widget doesn't give you a number to check, don't check its value.
        pass

    titr = make_titration()

    if sender == "b_tab_button":  # Everything but the buffer tab button
        # Perform bjerrum calculations
        bx = list(titr.ph)
        bys = [list(alpha) for alpha in titr.alpha_analyte.T]

        # For every alpha value list, plot the alpha values at every pH and add the line to the plot
        for num, alpha in enumerate(bys):
            add_line_series(plot="Main Plot", name=f"species{num}", x=bx, y=alpha, weight=2)

        # Relabel the plot, and x and y axes
        configure_item("Main Plot", label="Speciation Plot", x_axis_name="pH", y_axis_name="Relative Speciation")

    else:
        # Relabel the plot, and x and y axes
        configure_item("Main Plot", label="Titration", x_axis_name="Volume (mL)", y_axis_name="pH")

        # Perform titration calculations
        tx = list(titr.volume_titrant_t)
        ty = list(titr.ph_t)

        # plot the calculations
        add_line_series(
            plot="Main Plot",
            name="Titration Curve",
            x=tx,
            y=ty,
            weight=2,
            color=[0, 255, 255, 255],  # Cyan
        )

        if get_value("buffer_regions"):
            vols, pHs = titr.find_buffer_points()

            vols = list(vols)
            # It's impossible to have a pH > 14 in water, don't find the buffers for a solution which cannot exist.
            pHs = [x for x in pHs if x < 14]

            add_scatter_series(
                plot="Main Plot",
                name="Buffer Points",
                x=vols,
                y=pHs,
                fill=[255, 0, 0, 255],  # Red
                outline=[255, 0, 0, 255],  # Red
                weight=2,
            )

            # Add labels to the volumes of each point
            for vol, pH in zip(vols, pHs):
                if titr.analyte.acidic:
                    add_annotation("Main Plot", x=vol, y=pH, text=f"{vol:.5g} mL", xoffset=5, yoffset=5)
                else:
                    # Annotations need to be above the line if the solution is basic to prevent the line from clipping
                    add_annotation(
                        "Main Plot",
                        x=vol,
                        y=pH,
                        text=f"{vol:.5g} mL",
                        xoffset=5,
                        yoffset=-5,
                    )

        if get_value("equiv"):
            vols, pHs = titr.find_equiv_points()

            vols = list(vols)
            pHs = list(pHs)

            add_scatter_series(
                plot="Main Plot",
                name="Equivalence Points",
                x=vols,
                y=pHs,
                fill=[0, 255, 0, 255],  # Green
                outline=[0, 255, 0, 255],  # Green
                weight=2,
            )

            # Add labels to the volumes of each point
            for vol, pH in zip(vols, pHs):
                if titr.analyte.acidic:
                    add_annotation("Main Plot", x=vol, y=pH, text=f"{vol:.5g} mL", xoffset=5, yoffset=5)
                else:
                    # Annotations need to be above the line if the solution is basic to prevent the line from clipping
                    add_annotation(
                        "Main Plot",
                        x=vol,
                        y=pH,
                        text=f"{vol:.5g} mL",
                        xoffset=5,
                        yoffset=-5,
                    )

        if get_value("1stderiv"):
            volume, pHderiv = titr.deriv(degree=1)

            data = titr._scale_data(pHderiv, get_value("1dscaler"))

            add_line_series(
                plot="Main Plot",
                name="First Derivative",
                x=list(volume),
                y=list(data),
                weight=2,
                color=[255, 0, 255, 255],  # Purple
            )

        if get_value("2ndderiv"):
            volume, pHderiv = titr.deriv(degree=2)

            data = titr._scale_data(pHderiv, get_value("2dscaler"))

            add_line_series(
                plot="Main Plot",
                name="Second Derivative",
                x=list(volume),
                y=list(data),
                weight=2,
                color=[255, 255, 0, 255],  # Yellow
            )


def save_titr_data(sender, data):
    titr = make_titration()
    title = f"{get_value('aname')}_{get_value('tname')}_titration".replace(" ", "_")
    dw.write_titration_data(titr, title=title)


def save_bjer_data(sender, data):
    titr = make_titration()
    title = f"{get_value('aname')}_{get_value('tname')}_species".replace(" ", "_")
    dw.write_alpha_data(titr, title=title)


def save_ana_data(sender, data):
    titr = make_titration()
    title = f"{get_value('aname')}_{get_value('tname')}_analysis".replace(" ", "_")
    dw.write_analysis_data(titr, title=title)


# Main gui formatting
with window("Main Window", label="Something Else", autosize=True):
    set_main_window_size(width=1270, height=850)

    with tab_bar("Main Tab bar"):
        add_tab_button("t_tab_button", label="Titration", callback=plot_callback)
        add_tab_button("b_tab_button", label="Speciation", callback=plot_callback)

        add_same_line()
        add_text(" " * 127)  # TODO ask for right adjust text on DPG
        add_same_line()
        add_button("T-Builder by Jkelowitt", callback=open_link, tip="Open in github in browser")

        with group("Data Entry", width=data_width):
            add_text("Analyte Data")
            add_input_text(
                "aname",
                label="Name",
                default_value="Citric Acid",
                callback=plot_callback,
                tip="Enter the name of the analyte. This is used when making the data files.",
            )

            add_input_float(
                "aconc",
                label="Concentration (M)",
                default_value=0.10,
                callback=plot_callback,
                callback_data=[0],
                tip="Enter the concentration of the analyte in molarity.",
            )

            add_input_text(
                "apk",
                label="pKa value(s)",
                default_value="3.13, 4.76, 6.40",
                callback=plot_callback,
                tip="Enter the pKa values of the analyte. Separate them with commas if there are more than one.",
            )

            add_input_float(
                "avol",
                label="Volume (mL)",
                default_value=25,
                callback=plot_callback,
                callback_data=[0],
                tip="Enter the volume of the analyte in mL.",
            )

            add_checkbox(
                "aa",
                label="Acidic",
                default_value=True,
                callback=plot_callback,
                tip="Check this box if the analyte acts as an acid during this titration.",
            )  # TODO This may be automatable. Think "if pKa_a > pKa_t, then..."

            add_dummy(height=25)
            add_text("Titrant Data")

            add_input_text(
                "tname",
                label="Name",
                default_value="KOH",
                callback=plot_callback,
                tip="Enter the name of the titrant. This is used when naming the data files.",
            )

            add_input_float(
                "tconc",
                label="Concentration (M)",
                default_value=0.10,
                callback=plot_callback,
                callback_data=[0],
                tip="Enter the concentration of the titrant in molarity.",
            )

            add_input_text(
                "tpk",
                label="pKa value(s)",
                default_value="14.76",
                callback=plot_callback,
                tip="Enter the pKa values of the titrant. Separate them with commas if there are more than one.",
            )

            add_dummy(height=25)
            add_text("Titration Settings")

            add_input_int(
                "precision",
                label="Number of Points",
                default_value=2,
                callback=plot_callback,
                tip="The number of pH points to calculate. (10^n items)",
                width=65,
            )

            add_input_int(
                "temperature",
                label="Temperature (C)",
                default_value=25,
                callback=plot_callback,
                tip="The temperature at which the titration occurs. (0 - 100 C)",
                width=65,
            )

            add_dummy(height=25)
            add_text("Perform Titration Analysis")

            add_checkbox(
                "buffer_regions",
                label="Show Buffering Points",
                default_value=False,
                callback=plot_callback,
                tip="Show the center of the buffering regions on the Titration plot.",
            )

            add_checkbox(
                "equiv",
                label="Show Equivalence Points",
                default_value=False,
                callback=plot_callback,
                tip="Show the equivalence points on the Titration plot.",
            )

            add_checkbox(
                "1stderiv",
                label="Show normalized y'",
                default_value=False,
                callback=plot_callback,
                tip="Show the normalized 1st Derivative of the Titration plot",
            )

            add_checkbox(
                "2ndderiv",
                label="Show normalized y''",
                default_value=False,
                callback=plot_callback,
                tip="Show the normalized 2nd Derivative of the Titration plot.",
            )

            add_drag_float(
                "1dscaler",
                label="Scale y'",
                default_value=8,
                min_value=1,
                speed=0.1,
                width=80,
                format="%0.2f",
                callback=plot_callback,
                tip="Scale the 1st Derivative of the Titration plot.",
            )

            add_drag_float(
                "2dscaler",
                label="Scale y''",
                default_value=2,
                min_value=1,
                speed=0.1,
                width=80,
                format="%0.2f",
                callback=plot_callback,
                tip="Scale the 2nd Derivative of the Titration plot.",
            )

            add_dummy(height=25)
            add_text("Save Data to CSV")
            add_button("Save Titration Data ", callback=save_titr_data)
            add_button("Save Speciation Data", callback=save_bjer_data)
            add_button("Save Analysis Data", callback=save_ana_data)

        # Put the titration curve under the data entry section
        add_same_line()
        with group("TitrationPlotGroup"):
            add_plot(
                "Main Plot",
                label="Titration Curve",
                query_callback=query,
                width=plot_width,
                height=plot_height,
                anti_aliased=True,
                x_axis_name="Volume (ml)",
                y_axis_name="pH",
            )

    plot_callback("equiv", [])  # Make the plots appear on program start

# Run the curve.
if __name__ == "__main__":
    start_dearpygui(primary_window="Main Window")
